kong/utils.go

```diff
@@ -741,3 +824,247 @@ func FillPluginsDefaults(plugin *Plugin, schema Schema) error {
+				acceptNullValue := deprecatedFieldValue == nil
+				for _, path := range parseReplacedWithPaths(shorthandFieldConfig) {
+					clearUnmatchingDeprecationsForGivenPath(path, newPluginConfig, oldPluginConfig, acceptNullValue)
+				}
+			} else {
+				// The new plugin configuration does not contain deprecated fields.
+				// However, for backwards compatibility, Kong sends deprecated fields in the response.
+				// To ensure consistent diffs, we need to Delete deprecated fields from the old plugin configuration
+				// that Kong sent us.
+				delete(oldPluginConfig, deprecatedFieldName)
+			}
+		}
+
+		return true
+	})
+}
+
+// traverseConfigurationsAndExecute iterates over the fields in two plugin configurations (configA and configB),
+// and executes the provided function 'f' on nested fields of type "record" that exist in both configurations.
+//
+// Parameters:
+//
+//	configA - The first configuration to compare.
+//	configB - The second configuration to compare.
+//	schema - The schema describing the fields of the configurations.
+//	f - A function that will be executed on each pair of matching nested fields of type "record" in configA and configB.
+func traverseConfigurationsAndExecute(
+	configA Configuration,
+	configB Configuration,
+	schema *gjson.Result,
+	f func(Configuration, Configuration, *gjson.Result),
+) {
+	configFields := schema.Get("fields")
+	configFields.ForEach(func(_, value gjson.Result) bool {
+		field := value.Map()
+
+		for fieldName, fieldConfig := range field {
+			if fieldType := fieldConfig.Get("type"); fieldType.String() == "record" {
+				var nestedConfigA map[string]interface{}
+				if fieldA, ok := configA[fieldName].(map[string]interface{}); ok {
+					nestedConfigA = fieldA
+				}
+
+				var nestedConfigB map[string]interface{}
+				if fieldB, ok := configB[fieldName].(map[string]interface{}); ok {
+					nestedConfigB = fieldB
+				}
+
+				if nestedConfigA != nil && nestedConfigB != nil {
+					f(nestedConfigA, nestedConfigB, &fieldConfig)
+				}
+			}
+		}
+
+		return true
+	})
+}
+
+// same as ClearUnmatchingDeprecations but this function below is adjusted for recursive use (which is required
+// when the schema contains nested records and they can have nested shorthands).
+func clearUnmatchingDeprecationsHelper(
+	newPluginConfig Configuration,
+	oldPluginConfig Configuration,
+	schema *gjson.Result,
+) {
+	clearCurrentLevelUnmatchingDeprecations(newPluginConfig, oldPluginConfig, schema)
+
+	// Recursively walk through configuration to clear any nested unmatching deprecations.
+	traverseConfigurationsAndExecute(
+		newPluginConfig,
+		oldPluginConfig,
+		schema,
+		clearUnmatchingDeprecationsHelper,
+	)
+}
+
+// ClearUnmatchingDeprecations is a function that go through a pair of
+// configurations: newPlugin and oldPlugin, and by using schema it makes sure those two configurations
+// are aligned.
+// It does so by removing deprecated or new fields in "oldPlugin" that were not defined in "oldPlugin".
```

Comment:
> removing deprecated or new fields in "oldPlugin" that were not defined in "oldPlugin"

This is somewhat unclear. This removes things from `oldPlugin` which were not defined in `oldPlugin` 🤔 ?